lib/adapters/gun.ex

```diff
@@ -0,0 +1,183 @@
+      with [_protocol, _host, _port, _endpoint] = conn_params <- parse_url(url) do
+        case create_ws_connection(conn_params, timeout, args) do
+          {:connected, conn} ->
+            {:ok, %State{connection: conn, receiver: args[:message_receiver]}}
+
+          {:error, reason} ->
+            {:stop, reason}
+        end
+      else
+        {:error, reason} ->
+          {:stop, reason}
+      end
+    end
+
+    @impl GenServer
+    def handle_cast(:disconnect, %State{connection: conn, receiver: receiver} = state) do
+      :ok = :gun.close(conn)
+      notify_status(receiver, {:disconnected, "disconnected on request"})
+      {:stop, state}
+    end
+
+    def handle_cast({:send, frame}, %State{connection: conn} = state) do
+      :ok = :gun.ws_send(conn, {:text, frame})
+      {:noreply, state}
+    end
+
+    @impl GenServer
+    def handle_info(
+          {:gun_ws, _conn, _stream_ref, {:text, frame}},
+          %State{receiver: receiver} = state
+        ) do
+      forward_frame(receiver, frame)
+      {:noreply, state}
+    end
+
+    def handle_info(
+          {:gun_ws, _conn, _stream_ref, {:close, _, _}},
+          %State{receiver: receiver} = state
+        ) do
+      notify_status(receiver, {:disconnected, "remote disconnect"})
+      {:noreply, state}
+    end
+
+    def handle_info(
+          {:DOWN, _ref, :process, conn, reason},
+          %State{connection: conn, receiver: receiver} = state
+        ) do
+      notify_status(receiver, {:disconnected, reason})
+      {:stop, state}
+    end
+
+    # ignore protocol and try to connect without tsl
+    defp create_ws_connection([_protocol, host, port, path], timeout, %{
+           extra_headers: extra_headers
+         }) do
+      with {:ok, conn} <-
+             :gun.open(String.to_charlist(host), port, %{
+               transport: :tcp,
+               connect_timeout: timeout
+             }),
+           _ref <- Process.monitor(conn),
+           {:ok, _protocol} <- :gun.await_up(conn) do
+        protocol = parse_sec_protocol(extra_headers)
+
+        options =
+          unless is_nil(protocol) do
+            %{protocols: [protocol]}
+          else
+            %{}
+          end
+
+        :gun.ws_upgrade(conn, path, extra_headers, options)
+
+        receive do
+          {:gun_upgrade, conn, _stream_ref, ["websocket"], _headers} ->
+            {:connected, conn}
+
+          {:gun_response, _conn, _, _, _status, _} ->
+            {:error, :upgrade_failed}
+
+          {:gun_error, _conn, _, reason} ->
+            {:error, reason}
+
+          error ->
+            {:error, error}
+        end
+      else
+        {:error, _} = error ->
+          error
+      end
+    end
+
+    def parse_sec_protocol(headers) do
+      sec_protocols_keys = ["sec-websocket-protocol", "Sec-WebSocket-Protocol"]
+
+      protocol = headers |> Enum.find(fn {key, _val} -> key in sec_protocols_keys end)
+
+      with {_key, value} <- protocol do
```

Comment:
1. I'd rather use `case` or `if` here, it's not obvious this function may return  `nil`
2. You can move list generation here, it seems empty list is the default value and won't break anything https://ninenines.eu/docs/en/gun/2.0/manual/gun/

Reply:
I get `:badmatch` error on empty list so I guess point 2 can't be applied.

Reply:
This might be specific to 2.0 version :man_shrugging: